File: Libs/pollilib/python/tests/test_images_feeds.py
import os
import tempfile

from polliLib import PolliClient
from .conftest import FakeResponse, FakeSession


def test_generate_image_stream_to_file_and_fetch_bytes(tmp_path: tempfile.TemporaryDirectory):
    # Streamed image generation
    chunks = [b'a', b'b', b'c']
    fs = FakeSession()

    def fake_get(url, **kw):
        # For prompt URL with stream=True return chunks; otherwise bytes
        if kw.get('stream'):
            fs.last_get = (url, kw)
            return FakeResponse(content_chunks=chunks)
        fs.last_get = (url, kw)
        return FakeResponse(content=b'XYZ')

    fs.get = fake_get
    c = PolliClient(session=fs)
    out = c.generate_image("test", out_path=os.path.join(tmp_path, "gen.jpg"))
    assert os.path.exists(out)
    _, params = fs.last_get
    assert params["params"]["safe"] == "false"
    with open(out, 'rb') as f:
        assert f.read() == b''.join(chunks)

    # Direct fetch
    data = c.fetch_image("http://image/url.jpg")
    assert data == b'XYZ'


def test_save_image_timestamped(tmp_path: tempfile.TemporaryDirectory):
    fs = FakeSession()
    fs.get = lambda url, **kw: FakeResponse(content=b'JPG')
    c = PolliClient(session=fs)
    path = c.save_image_timestamped("test", images_dir=str(tmp_path))
    assert os.path.exists(path)
    assert path.endswith('.jpeg')


def test_image_feed_stream_parsing_and_bytes_and_data_url():
    # Without include options
    lines = [
        'data: {"prompt":"p1","imageURL":"http://img/1.jpg","model":"flux","seed":123}',
        'data: [DONE]'
    ]
    fs = FakeSession()
    fs.get = lambda url, **kw: FakeResponse(stream_lines=lines)
    c = PolliClient(session=fs)
    events = list(c.image_feed_stream())
    assert events and events[0]['prompt'] == 'p1'

    # include_bytes
    def get_with_image(url, **kw):
        if url.endswith('/feed'):
            return FakeResponse(stream_lines=lines)
        return FakeResponse(content=b'\xff\xd8\xff', headers={'Content-Type': 'image/jpeg'})

    fs = FakeSession()
    fs.get = get_with_image
    c = PolliClient(session=fs)
    events = list(c.image_feed_stream(include_bytes=True))
    assert events and events[0]['image_bytes'] == b'\xff\xd8\xff'

    # include_data_url
    fs = FakeSession()
    fs.get = get_with_image
    c = PolliClient(session=fs)
    events = list(c.image_feed_stream(include_data_url=True))
    assert events and events[0]['image_data_url'].startswith('data:image/jpeg;base64,')


def test_text_feed_stream_parsing():
    lines = [
        'data: {"model":"openai","messages":[{"role":"user","content":"hi"}],"response":"Hello"}',
        'data: [DONE]'
    ]
    fs = FakeSession()
    fs.get = lambda url, **kw: FakeResponse(stream_lines=lines)
    c = PolliClient(session=fs)
    ev = next(iter(c.text_feed_stream()))
    assert ev['model'] == 'openai' and ev['response'] == 'Hello'

File: Libs/pollilib/javascript/tests/test_images_feeds.js
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';

import { PolliClient } from '../polliLib/index.js';
import { FakeResponse, SeqFetch } from './helpers.js';

test('generate_image streams to file and fetch_image returns bytes', async (t) => {
  const tmpDir = await fs.promises.mkdtemp(path.join(process.cwd(), 'tmp-img-'));
  const outPath = path.join(tmpDir, 'gen.jpg');
  const seq = new SeqFetch([
    new FakeResponse({ content: Buffer.from('abc') }),
  ]);
  // For streaming path: provide body stream lines as bytes and capture calls
  const calls = [];
  seq.fetch = async (url, opts = {}) => {
    calls.push({ url: String(url), opts });
    if (opts?.signal && opts.method === 'GET' && String(url).includes('/prompt/')) {
      // Simulate streaming chunks
      const lines = ['a', 'b', 'c'];
      return new FakeResponse({ streamLines: lines });
    }
    return new FakeResponse({ content: Buffer.from('XYZ') });
  };
  const c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const saved = await c.generate_image('test', { outPath });
  assert.equal(saved, outPath);
  const promptCall = calls.find((call) => String(call.url).includes('/prompt/'));
  const promptUrl = new URL(promptCall.url);
  assert.equal(promptUrl.searchParams.get('safe'), 'false');
  const data = await fs.promises.readFile(outPath);
  // Our fake stream writes chunks with newlines per line; normalize
  assert.equal(data.toString('utf-8').replace(/\n/g, ''), 'abc');

  const bytes = await c.fetch_image('http://x/y.jpg');
  assert.equal(Buffer.isBuffer(bytes), true);
});

test('save_image_timestamped creates file', async () => {
  const tmpDir = await fs.promises.mkdtemp(path.join(process.cwd(), 'tmp-img-'));
  const seq = new SeqFetch([ new FakeResponse({ content: Buffer.from('JPG') }) ]);
  const c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const p = await c.save_image_timestamped('p', { imagesDir: tmpDir });
  assert.ok(fs.existsSync(p));
  assert.ok(p.endsWith('.jpeg'));
});

test('image_feed_stream parses and can include bytes/data URL', async () => {
  const lines = [
    'data: {"prompt":"p1","imageURL":"http://img/1.jpg","model":"flux","seed":123}',
    'data: [DONE]'
  ];
  // Case 1: parse only
  let seq = new SeqFetch([ new FakeResponse({ streamLines: lines }) ]);
  let c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const ev1 = [];
  for await (const ev of c.image_feed_stream()) ev1.push(ev);
  assert.equal(ev1[0].prompt, 'p1');

  // Case 2: include bytes
  seq = new SeqFetch();
  seq.fetch = async (url, opts = {}) => {
    if (String(url).endsWith('/feed')) return new FakeResponse({ streamLines: lines });
    return new FakeResponse({ content: Buffer.from([0xFF,0xD8,0xFF]), headers: { 'Content-Type': 'image/jpeg' } });
  };
  c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const ev2 = [];
  for await (const ev of c.image_feed_stream({ includeBytes: true })) ev2.push(ev);
  assert.equal(Buffer.compare(ev2[0].image_bytes, Buffer.from([0xFF,0xD8,0xFF])), 0);

  // Case 3: include data URL
  seq = new SeqFetch();
  seq.fetch = async (url, opts = {}) => {
    if (String(url).endsWith('/feed')) return new FakeResponse({ streamLines: lines });
    return new FakeResponse({ content: Buffer.from([0xFF,0xD8,0xFF]), headers: { 'Content-Type': 'image/jpeg' } });
  };
  c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const ev3 = [];
  for await (const ev of c.image_feed_stream({ includeDataUrl: true })) ev3.push(ev);
  assert.ok(ev3[0].image_data_url.startsWith('data:image/jpeg;base64,'));
});

test('text_feed_stream parses events', async () => {
  const lines = [
    'data: {"model":"openai","messages":[{"role":"user","content":"hi"}],"response":"Hello"}',
    'data: [DONE]'
  ];
  const seq = new SeqFetch([ new FakeResponse({ streamLines: lines }) ]);
  const c = new PolliClient({ fetch: seq.fetch.bind(seq) });
  const it = c.text_feed_stream();
  const first = await (async () => { for await (const ev of it) return ev; })();
  assert.equal(first.model, 'openai');
  assert.equal(first.response, 'Hello');
});
